src/commands/force/mdapi/beta/retrieve.ts

```diff
@@ -0,0 +1,304 @@
+    }),
+    apiversion: flags.builtin({
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore force char override for backward compat
+      char: 'a',
+      description: messages.getMessage('flags.apiversion'),
+      longDescription: messages.getMessage('flagsLong.apiversion'),
+    }),
+    verbose: flags.builtin({
+      description: messages.getMessage('flags.verbose'),
+      longDescription: messages.getMessage('flagsLong.verbose'),
+    }),
+  };
+
+  protected retrieveResult: RetrieveResult;
+  private sourceDir: string;
+  private retrieveTargetDir: string;
+  private zipFileName: string;
+  private unzip: boolean;
+  private wait: Duration;
+  private isAsync: boolean;
+  private mdapiRetrieve: MetadataApiRetrieve;
+
+  public async run(): Promise<RetrieveCommandResult | RetrieveCommandAsyncResult> {
+    await this.retrieve();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  protected async retrieve(): Promise<void> {
+    const packagenames = this.getFlag<string[]>('packagenames');
+    if (!packagenames) {
+      this.sourceDir = this.resolveRootDir(this.getFlag<string>('sourcedir'));
+    }
+    this.retrieveTargetDir = this.resolveOutputDir(this.getFlag<string>('retrievetargetdir'));
+    const manifest = this.resolveManifest(this.getFlag<string>('unpackaged'));
+    const singlePackage = this.getFlag<boolean>('singlepackage');
+    this.zipFileName = this.getFlag<string>('zipfilename') || 'unpackaged.zip';
+    this.unzip = this.getFlag<boolean>('unzip');
+    this.wait = this.getFlag<Duration>('wait');
+    this.isAsync = this.wait.quantity === 0;
+
+    if (singlePackage && packagenames?.length > 1) {
+      throw SfdxError.create('@salesforce/plugin-source', 'md.retrieve', 'InvalidPackageNames', [
+        packagenames.toString(),
+      ]);
+    }
+
+    this.ux.startSpinner(spinnerMessages.getMessage('retrieve.main', [this.org.getUsername()]));
+    this.ux.setSpinnerStatus(spinnerMessages.getMessage('retrieve.componentSetBuild'));
+
+    this.componentSet = await ComponentSetBuilder.build({
+      apiversion: this.getFlag<string>('apiversion'),
+      packagenames,
+      sourcepath: this.sourceDir ? [this.sourceDir] : undefined,
+      manifest: manifest && {
+        manifestPath: manifest,
+        directoryPaths: [],
+      },
+    });
+
+    await this.lifecycle.emit('preretrieve', { packageXmlPath: manifest });
+
+    this.ux.setSpinnerStatus(spinnerMessages.getMessage('retrieve.sendingRequest', [this.componentSet.apiVersion]));
+
+    this.mdapiRetrieve = await this.componentSet.retrieve({
+      usernameOrConnection: this.org.getUsername(),
+      output: this.retrieveTargetDir,
+      packageOptions: this.getFlag<string[]>('packagenames'),
+      format: 'metadata',
+      singlePackage,
+      zipFileName: this.zipFileName,
+      unzip: this.unzip,
+    });
+
+    Stash.set('MDAPI_RETRIEVE', {
```

Comment:
Good call, I'll update.

Reply:
TypeScript resolution didn't like that so I'm keeping it as is rather than coercing like this:
`Stash.set(Stash.KEYS.MDAPI_RETRIEVE as StashKey, { ... });`